rest/handler.go

```diff
@@ -400,38 +407,61 @@ func (h *handler) checkAuth(context *db.DatabaseContext) (err error) {
 	return nil
 }
 
-func (h *handler) checkAdminAuth(dbContext *db.DatabaseContext, checkPermissions []string, storePermissionsResults bool) error {
+func checkAdminAuth(serverContext *ServerContext, dbContext *db.DatabaseContext, basicAuthUsername, basicAuthPassword string, checkPermissions []string) (map[string]bool, error) {
 	var managementEndpoints []string
 	var err error
 	var httpClient *http.Client
 
 	if dbContext != nil {
 		managementEndpoints, httpClient, err = dbContext.ObtainManagementEndpointsAndHTTPClient()
 		if err != nil {
-			return base.HTTPErrorf(http.StatusInternalServerError, "")
+			return nil, base.HTTPErrorf(http.StatusInternalServerError, "")
 		}
 	} else {
-		managementEndpoints, httpClient, err = h.server.ObtainManagementEndpointsAndHTTPClient()
+		managementEndpoints, httpClient, err = serverContext.ObtainManagementEndpointsAndHTTPClient()
 		if err != nil {
-			return base.HTTPErrorf(http.StatusInternalServerError, "")
+			return nil, base.HTTPErrorf(http.StatusInternalServerError, "")
 		}
 	}
 
-	username, password := h.getBasicAuth()
-	statusCode, permResults, err := h.server.CheckPermissions(httpClient, managementEndpoints, username, password, checkPermissions)
+	statusCode, permResults, err := CheckPermissions(httpClient, managementEndpoints, basicAuthUsername, basicAuthPassword, checkPermissions)
 	if err != nil {
-		return base.HTTPErrorf(http.StatusInternalServerError, "")
+		return nil, base.HTTPErrorf(http.StatusInternalServerError, "")
+	}
+
+	if statusCode == http.StatusOK {
+		return permResults, nil
+	}
+
+	// If status code was not 'ok' or 'forbidden' return an error
+	// If forbidden user has authenticated correctly but is not authorized via permissions. We'll fall through to try
+	// with roles.
+	if statusCode != http.StatusForbidden {
+		return nil, base.HTTPErrorf(statusCode, "")
 	}
 
-	if storePermissionsResults {
-		h.permissionsResults = permResults
+	var requestRoles []string
+	var bucketName string
+	if dbContext != nil {
+		requestRoles = []string{"mobile_sync_gateway", "bucket_full_access", "admin"}
+		bucketName = dbContext.Bucket.GetName()
```

Comment:
If you do remove serverContext/dbContext above, passing down a role scope seems reasonable to me, where `scope:"" == cluster`

Reply:
Agree with Ben's comments here and above - instead of passing in the entire serverContext/dbContext, make this function more stand-alone and pass in only what's needed by the function.  Makes future refactoring of serverContext/dbContext easier, as we know exactly what parts of the two this function is depending on.

Reply:
Not sure how keen I am on it given the management endpoint work has to be split out so its not all encapsulated into one function anymore but I understand the benefits so done.